helix-term/src/grammars.rs

```diff
@@ -0,0 +1,290 @@
+        if !grammar_dir.join(".git").is_dir() {
+            Command::new("git")
+                .args(["init"])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Could not execute 'git'");
+        }
+
+        // ensure the remote matches the configured remote
+        if get_repository_info(&grammar_dir, vec!["remote", "get-url", REMOTE_NAME])
+            != Some(remote.clone())
+        {
+            set_remote(&grammar_dir, &remote);
+        }
+
+        // ensure the revision matches the configured revision
+        if get_repository_info(&grammar_dir, vec!["rev-parse", "HEAD"]) != Some(revision.clone()) {
+            // Fetch the exact revision from the remote.
+            // Supported by server-side git since v2.5.0 (July 2015),
+            // enabled by default on major git hosts.
+            Command::new("git")
+                .args(["fetch", REMOTE_NAME, &revision])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Failed to execute 'git'");
+
+            Command::new("git")
+                .args(["checkout", &revision])
+                .current_dir(grammar_dir)
+                .output()
+                .expect("Failed to execute 'git'");
+
+            println!(
+                "Grammar '{}' checked out at '{}'.",
+                grammar.grammar_id, revision
+            );
+        } else {
+            println!("Grammar '{}' is already up to date.", grammar.grammar_id);
+        }
+    };
+}
+
+// Sets the remote for a repository to the given URL, creating the remote if
+// it does not yet exist.
+fn set_remote(repository: &Path, remote_url: &str) {
+    if !Command::new("git")
+        .args(["remote", "set-url", REMOTE_NAME, remote_url])
+        .current_dir(repository)
+        .output()
+        .expect("Failed to execute 'git'")
+        .status
+        .success()
+        && !Command::new("git")
+            .args(["remote", "add", REMOTE_NAME, remote_url])
+            .current_dir(repository)
+            .output()
+            .expect("Failed to execute 'git'")
+            .status
+            .success()
+    {
+        eprintln!("Failed to set remote '{}'", remote_url);
+    }
+}
+
+fn get_repository_info(repository: &Path, args: Vec<&str>) -> Option<String> {
+    let output = Command::new("git")
+        .args(args)
+        .current_dir(repository)
+        .output()
+        .expect("Failed to execute 'git'");
+    if output.status.success() {
+        let mut remote = String::from_utf8_lossy(output.stdout.as_slice()).into_owned();
+        // remove trailing newline
+        remote.pop();
+        Some(remote)
+    } else {
+        None
+    }
+}
+
+fn build_grammar(grammar: GrammarConfiguration) {
+    let grammar_dir = if let GrammarSource::Local { ref path } = grammar.source {
+        PathBuf::from(path)
+    } else {
+        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
```

Comment:
This is an absolute path of the *build machine's* directory of the source code. It will not work on anyone else's machine. This should be `HELIX_RUNTIME`.